### PR TITLE
Add swift 6 support

### PR DESCRIPTION
**Pull request by petar-jadek**
## Summary

Added Swift 6 support 
**Related issue**:  [#9](https://github.com/infinum/ios-collar/issues/9) 

## Changes

Updated code to work on Swift 6 compiler. Updated issues due to nested objects throwing error on serialization. No new features or capabilities were introduced.

### Type

- [ ] **Feature**: This pull request introduces a new feature.
- [x] **Bug fix**: This pull request fixes a bug.
- [x] **Refactor**: This pull request refactors existing code.
- [x] **Documentation**: This pull request updates documentation.
- [ ] **Other**: This pull request makes other changes.

#### Additional information

- [x] This pull request introduces a **breaking change**.

### Description

Added `LoggerJsonValue` and changed the serialization so nested objects can be serialized.

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have tested my changes, including edge cases.
- [x] I have added necessary tests for the changes introduced (if applicable).
- [x] I have updated the documentation to reflect my changes (if applicable).

## Additional notes

-

### CODE REVIEW

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift [hunk 1]
```diff
@@ -8,134 +8,66 @@
...
+            self.postUpdateNotification()
+        }
     }
 
     public func clearLog(_ logItem: LogItem) {
-        logs = logs.filter { $0.id != logItem.id }
+        queue.async(flags: .barrier) {
+            self._logs.removeAll { $0.id == logItem.id }
+            self.postUpdateNotification()
+        }
```
Comment (kjaklinovic) [inline Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift]: Why don't you just use serial queue since you are always using this with a barrier flag ?

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift [hunk 2]
```diff
@@ -8,134 +8,66 @@
...
-    public private(set) var logs: [LogItem] = [] {
-        didSet {
-            NotificationCenter.default.post(AnalyticsCollectionManager.Notification.didUpdateLogs)
-        }
+
+    private let queue = DispatchQueue(label: "analytics.logs.queue", attributes: .concurrent)
+    private var _logs: [LogItem] = []
+
+    public var logs: [LogItem] {
+        queue.sync { _logs }
```
Comment (kjaklinovic) [inline Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift]: I don't really like to use sync. Maybe some documentation is needed here, like don't call this from the main thread or something similar since this will suspend that thread.

Comment (kjaklinovic) [reply Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift]: How hard would it be to make this logs async operation, combine, async await, rx, completion, whatever you have ready on the project.

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift [hunk 3]
```diff
@@ -8,134 +8,66 @@
...
 
-    var pasteboardString: String {
-        let parameters = "Parameters: " + (subtitleDisplay ?? "")
-        let timestamp = "Timestamp: " + timestamp.description
-        return type.rawValue + ": " + name + "\n" + timestamp + "\n" + parameters
+    private func postUpdateNotification() {
+        DispatchQueue.main.async {
+            NotificationCenter.default.post(AnalyticsCollectionManager.Notification.didUpdateLogs)
+        }
     }
```
Comment (kjaklinovic) [inline Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift:112]: My preference here would be to just post the notification, whichever thread it is and add documentation that this needs to be observed on the main thread on the call site if used to refresh ui or something.

File: Sources/Collar/Classes/Collar/LoggerJsonValue.swift [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+extension LoggerJsonValue {
+    var jsonCompatible: Any {
+        switch self {
+        case .string(let s): return s
+        case .int(let i): return i
+        case .double(let d): return d
+        case .bool(let b): return b
+        case .array(let arr): return arr.map { $0.jsonCompatible }
+        case .object(let dict): return dict.mapValues { $0.jsonCompatible }
+        case .null: return NSNull()
```
Comment (kjaklinovic) [inline Sources/Collar/Classes/Collar/LoggerJsonValue.swift]: Omit return if that is your style.

File: Sources/Collar/Classes/Collar/LogItem.swift [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    public let name: String
+    public let timestamp: Date
+    public let value: String?
+    public let parameters: [String: LoggerJsonValue]?
+
+    private static var dateFormatter: ISO8601DateFormatter {
+        let formatter = ISO8601DateFormatter()
+        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
+        return formatter
+    }
```
Comment (Truba) [inline Sources/Collar/Classes/Collar/LogItem.swift]: This is just a static getter, correct?
It will recreate the dateFormatter on every call?
It's usually recommended to cache the date formatter, since creating a new one can be an expensive operation.
Something like

```suggestion
    private static let dateFormatter: ISO8601DateFormatter =  {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()
```

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift (hunk 2 above)
Comment (Truba) [reply Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift]: Since we are increasing the major version, I would say braking changes are ok 👍

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift (hunk 3 above)
Comment (Truba) [reply Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift:112]: Agree with "just post the notification".
But for the documentation, I would just mention that it's posted on the internal queue.

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift [hunk 6]
```diff
@@ -8,134 +8,66 @@
...
 
-    var subtitleDisplay: String? {
-        switch type {
-        case .event:
-            return paramsJSONString
-        case .userProperty, .screen:
-            return value
+    // MARK: - Private helpers
+
+    private func appendLog(_ log: LogItem) {
```
Comment (Truba) [inline Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift:105]: Why are you using concurrent queue if you are constantly barriering it?
Would it be simpler to just make ti serial?

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift (hunk 2 above)
Comment (Truba) [reply Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift]: Maybe updating `AnalyticsCollectionManager` to actor 🤔

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift [hunk 7]
```diff
@@ -8,134 +8,91 @@
...
-    init(type: LogType, name: String, timestamp: Date = Date(), value: String? = nil, parameters: [String: Any]? = nil) {
-        self.type = type
-        self.name = name
-        self.timestamp = timestamp
-        self.value = value
-        self.parameters = parameters
-    }
+/// Actor-based analytics collection manager providing thread-safe log management.
+/// All methods are async and must be awaited. Actor isolation provides compiler-verified thread safety.
+public actor AnalyticsCollectionManager {
```
Comment (kjaklinovic) [inline Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift]: I would say this is an overkill. 
In general await for analytics seems weird, I would put this into the same category as logging, i would prefer this is a fire and forget thing that is managed internally for thread safety. Having an async api creates some overhead, it's hard to interact with that since it's required you access it within the context of either Task or async function.

**Review by kjaklinovic — CHANGES_REQUESTED** [review]

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift [hunk 8]
```diff
@@ -8,50 +8,47 @@
...
 
+    private let queue = DispatchQueue(label: "com.infinum.collar.analytics", qos: .utility)
     private var _logs: [LogItem] = []
 
-    /// All collected log items.
-    /// - Returns: Array of log items in chronological order
-    /// - Note: This is an async operation that suspends until the actor can provide safe access
+    private init() {}
+
+    /// All collected log items. Synchronously waits for any pending writes to complete.
```
Comment (kjaklinovic) [inline Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift]: Is this a good idea, why don't we make fetching of all logs async ? 
Does that make a difference somewhere ? If this is called from the main thread, which it probably is, this will block the main thread. I would consider making this function async by either using async await or maybe just provide a completion block and se different wrappers for it where needed.

File: Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift [hunk 9]
```diff
@@ -8,50 +8,47 @@
...
     }
 
     /// Removes a specific log item.
     /// - Parameter logItem: The log item to remove
-    /// - Note: This method is async and must be awaited
-    public func clearLog(_ logItem: LogItem) async {
-        _logs.removeAll { $0.id == logItem.id }
-        postUpdateNotification()
+    public func clearLog(_ logItem: LogItem) {
+        queue.async { [self] in
```
Comment (kjaklinovic) [inline Sources/Collar/Classes/Collar/AnalyticsCollectionManager.swift]: Why not capturing self as weak in all those async closures  ?

File: Sources/Collar/Classes/Collar/LogItem.swift [hunk 10]
```diff
@@ -52,7 +46,9 @@ public struct LogItem: CustomStringConvertible, Identifiable, Sendable {
         var lines: [String] = []
         lines.append("Type: \(type.rawValue)")
         lines.append("Name: \(name)")
-        lines.append("Timestamp: \(LogItem.dateFormatter.string(from: timestamp))")
+        let formatter = ISO8601DateFormatter()
```
Comment (kjaklinovic) [inline Sources/Collar/Classes/Collar/LogItem.swift]: Date formatters are expensive to init. Maybe consider moving that to some type of a lazy loading (static)

File: Sources/Collar/Classes/List/SwiftUI/LogListView.swift [hunk 11]
```diff
@@ -121,12 +121,7 @@ struct LogListView: View {
     }
 
     private func updateLogs() {
-        Task {
-            let logs = await analyticsManager.logs
-            await MainActor.run {
-                items = logs.sorted(by: { $0.timestamp > $1.timestamp })
-            }
-        }
+        items = analyticsManager.logs.sorted(by: { $0.timestamp > $1.timestamp })
```
Comment (kjaklinovic) [inline Sources/Collar/Classes/List/SwiftUI/LogListView.swift]: This will block the main thread since you are calling logs here which iirc uses sync.
